### DISCUSSION TITLE
Google Plugin Support

### DISCUSSION THREAD

**Discussion by sirius-chao** [cefsharp/CefSharp#4937] [General]
In earlier versions of CefSharp, there was an IExtensionHandler for extending Chrome extensions. In the new version, IExtensionHandler has been removed. Does this mean that extension support is no longer available, or is there a new method for this? Please let me know, thank you.

**Comment by mitchcapper** [comment]
It means that with the alloy runtime gone there is no longer the extension loading support.  Alloy style does not support it but chrome style does work with extensions and as of 126 there is a devtools load extension call that is pretty sexy: https://chromedevtools.github.io/devtools-protocol/tot/Extensions/#method-loadUnpacked

It should have much better extension support now.

**Reply by amaitland** [reply]
Loading `chrome://extensions/` and enabling `Developer Mode` should also allow you to load unpackaged extensions

Example of switch to using `Chrome Styled` browsers using https://github.com/cefsharp/CefSharp/issues/4835#issuecomment-2182373137

Only supported in `WinForms` and `Wpf.HwndHost`

![image](https://github.com/user-attachments/assets/56ebd8e9-8046-4775-8cbb-837fed8dab91)

**Reply by mitchcapper** [reply]
and if you want to do it programmatically there is `context.SetPreference("extensions.ui.developer_mode",true`

**Reply by sirius-chao** [reply]
Thanks to both of you for your replies, @amaitland @mitchcapper 

I have successfully opened the extension page.

![image](https://github.com/user-attachments/assets/b96d631b-2731-414a-97dc-bc96280018bc)

But an exception is thrown when this extension is loaded, as shown in the figure below:

![image](https://github.com/user-attachments/assets/47d8d4f7-8ae8-4236-91fa-b67bc17d13ab)

I attempted to add settings.CefCommandLineArgs.Add("enable-experimental-extension-apis", "1"); in CefSettings, but unfortunately, it didn’t have the desired effect. Are there any additional parameters that need to be configured, or other potential solutions to this issue? I would appreciate your insights. Thank you!

**Comment by mitchcapper** [comment]
OK so the following are required:
*) Running v128 or newer
*) Use chrome runtime (`CefSharpSettings.RuntimeStyle = CefRuntimeStyle.Chrome;` before any initalization)
*) Be running WPF.HwndHost or WinForms
*) per https://chromedevtools.github.io/devtools-protocol/tot/Extensions/#method-loadUnpacked  `--enable-unsafe-extension-debugging` cef command line arg.  I *think*  amaitland  automatically has the docs on the c# generated classes as well.
*) unsafe extension loading enabled (manually in extensions or `context.SetPreference("extensions.ui.developer_mode",true);`)

Now it may be you still get that error.   I did also try to set the entire raw preference blob in a `IDictionary<string, object>`:   (note you would need to compute the chrome hash that is based on the file path but this isn't too hard code at bottom but just take a known good json of the pref for testing anyway)

```csharp
var json = @"{
	""oejmofjkmijelromghaoncbbecjiial"": {
		""active_permissions"": {
			""api"": [
				""storage""
			],
			""explicit_host"": [],
			""manifest_permissions"": [],
			""scriptable_host"": [
				""file:///*"",
				""http://*/*"",
				""https://*/*""
			]
		},
		""commands"": {},
		""content_settings"": [],
		""creation_flags"": 38,
		""events"": [],
		""first_install_time"": ""13372343007931350"",
		""from_webstore"": false,
		""granted_permissions"": {
			""api"": [
				""storage""
			],
			""explicit_host"": [],
			""manifest_permissions"": [],
			""scriptable_host"": [
				""file:///*"",
				""http://*/*"",
				""https://*/*""
			]
		},
		""incognito_content_settings"": [],
		""incognito_preferences"": {},
		""last_update_time"": ""13372343007931350"",
		""location"": 4,
		""newAllowFileAccess"": true,
		""path"": ""F:\\my\\extension"",
		""preferences"": {},
		""regular_only_preferences"": {},
		""state"": 1,
		""was_installed_by_default"": false,
		""was_installed_by_oem"": false,
		""withholding_permissions"": false
	}
}";
var jObj = JsonSerializer.Deserialize<Dictionary<string, object>>(json);
 context.SetPreference("extensions.settings", jStr, out var errSet);

``` 
It does give an args error but I haven't had time to track down if that is a CEF issue, a CEFSharp thing, or just totally not allowed by chrome.  I did try to set the extension id directly but get key now found in those cases.  At first I thought CEFSharp wasn't recursively serializing the dictionaries but looking at the code I don't think that is it.  I really need to modify the prefs.html cefclient test to take a dictionary and see if that also errors.  I haven't found a way to test devtools protocol with the cefclient to see about the loadExtension call.



```csharp
path = char.ToUpper(path[0]) + path.Substring(1);           
var bytes = Encoding.Unicode.GetBytes(path);
using var hash = SHA256.Create();
var hashed_bytes = hash.ComputeHash(bytes);
var hexDigits = "abcdefghijklmnop".ToCharArray();
var extID = String.Join("",hashed_bytes.Take(16).Select(b=> $"{hexDigits[b >> 4]}{hexDigits[b & 0xF]}"));
```

**Comment by mitchcapper** [comment]
OK! so not quite working in cefsharp but confirmed it works with the preference setting in cefclient to synthetically load an extension without other commands.    Granted cefclient must be restarted and seems to crash for the first close but it is added.  More things to debug:)

In terms of CEFSharp I have learned to avoid microsofts json solution like the plague did not expect it to serialize json to a string/object dictionary so oddly.   CEFSharp is still doing some awkward things with complex serialization objects that I need to debug as at greater depths it does literal 'key' 'value' sets.  Expando should be fine going to the dictionary but something is screwed up.  Just doing a call to get preference / set preference like:

```csharp
var curExtSettings = (Dictionary<string,object>) context.GetPreference("extensions.settings");
context.SetPreference("extensions.settings", curExtSettings, out errSet);
```
results in the prefs store being corrupted to (oddly chrome even stores this to disk):
```json
{
	"owoawenfoawenfoanwefonwaefe": [
		{
			"Key": "active_permissions",
			"Value": [
				{
					"Key": "api",
					"Value": [
						"management",
						"system.display",
						"system.storage",
						"webstorePrivate",
						"system.cpu",
						"system.memory",
						"system.network"
					]
				},
				{
					"Key": "explicit_host",
					"Value": [
						
					]
				},
				{
					"Key": "manifest_permissions",
					"Value": [
						
					]
				},
				{
					"Key": "scriptable_host",
					"Value": [
						
					]
				}
			]
```

A temporary fix as simple as converting it to a literal dictionary<string,object> recursively would probably work but I can probably get the CEFSharp serializer to handle it better.

**Comment by mitchcapper** [comment]
Ok somewhat obvious the crash on exit was due to it trying to cleanup the extension and use its lock which doesnt exist as it wasnt properly loaded:
```
 	[Inline Frame] libcef.dll!base::internal::LockImpl::Try() Line 103	C++
 	[Inline Frame] libcef.dll!base::internal::LockImpl::Lock() Line 94	C++
 	[Inline Frame] libcef.dll!base::Lock::Acquire(base::subtle::LockTracking tracking) Line 36	C++
 	[Inline Frame] libcef.dll!base::internal::BasicAutoLock<base::Lock>::BasicAutoLock(base::Lock & lock, base::subtle::LockTracking tracking) Line 147	C++
>	libcef.dll!extensions::ContentSettingsStore::ClearContentSettingsForExtension(const std::__Cr::basic_string<char,std::__Cr::char_traits<char>,std::__Cr::allocator<char>> & ext_id, extensions::api::types::ChromeSettingScope scope) Line 324	C++
 	libcef.dll!extensions::PreferenceAPI::ClearIncognitoSessionOnlyContentSettings() Line 419	C++
 	libcef.dll!extensions::PreferenceAPI::Shutdown() Line 373	C++
 	[Inline Frame] libcef.dll!DependencyManager::ShutdownFactoriesInOrder(void * context, const std::__Cr::vector<base::raw_ptr<KeyedServiceBaseFactory,1>,std::__Cr::allocator<base::raw_ptr<KeyedServiceBaseFactory,1>>> & factories) Line 204	C++
 	libcef.dll!DependencyManager::PerformInterlockedTwoPhaseShutdown(DependencyManager * dependency_manager1, void * context1, DependencyManager * dependency_manager2, void * context2) Line 175	C++
 	libcef.dll!ProfileImpl::~ProfileImpl() Line 971	C++
 	libcef.dll!ProfileImpl::~ProfileImpl() Line 913	C++
 	[Inline Frame] libcef.dll!std::__Cr::default_delete<Profile>::operator()(Profile * __ptr) Line 68	C++
 	[Inline Frame] libcef.dll!std::__Cr::unique_ptr<Profile,std::__Cr::default_delete<Profile>>::reset(Profile * __p) Line 280	C++
 	libcef.dll!ProfileDestroyer::DestroyOriginalProfileNow(std::__Cr::unique_ptr<Profile,std::__Cr::default_delete<Profile>> profile) Line 305	C++
 	libcef.dll!OriginalProfileDestroyer::DoDestroyUnderlyingProfile() Line 106	C++
 	[Inline Frame] libcef.dll!ProfileDestroyer::Timeout() Line 437	C++
 	libcef.dll!ProfileDestroyer::Start(const std::__Cr::set<content::RenderProcessHost *,std::__Cr::less<content::RenderProcessHost *>,std::__Cr::allocator<content::RenderProcessHost *>> & hosts) Line 0	C++
```
It actually does a null check for the lock but only as a debug assert not to avoid trying it.  This might mean that setting the preference in 
```csharp
protected override void OnRequestContextInitialized(IRequestContext requestContext)
{
    var curExtSettings = (Dictionary<string,object>)  requestContext.GetPreference("extensions.settings");
    curExtSettings["asdfasdfasdfasdf"] = Newtonsoft.Json.JsonConvert.DeserializeObject<IDictionary<string, object>>(extJson);
    if (! requestContext.SetPreference("extensions.settings",curExtSettings, out var err))
        throw new System.Exception(err);

    base.OnRequestContextInitialized(requestContext);
}
```
might just work to get it to load on first run.  I tried some initial testing but not quite yet.

**Reply by max3163** [reply]
When I use the Newtonsoft.Json.JsonConvert.DeserializeObject<IDictionary<string, object>>(extJson) to convert my string into a setting object it doesn't work properly. In fact it work only for first level and not deep object :

Example in the preference files after set it : 

```
	"jifbnihciifbfeiiijegkfnbigagacjk": {
				"ack_external": true,
				"active_bit": false,
				"active_permissions": [
					[
						[
							[],
							[],
							[],
							[]
						]
					],
					[
						[
							[]
						]
					],
					[
						[]
					],
					[
						[]
					]
				],
....
```

Can you help me ?

**Reply by max3163** [reply]
Ok, data are correct if I use this typing
```
curExtSettings[extId] = Newtonsoft.Json.JsonConvert.DeserializeObject<ExpandoObject>(json); 
```
But the extension is still not activated ...

**Comment by max3163** [comment]
I'm trying to add a plugin extension too and I follow your discussion to help me. 
With the chrome://extensions page I can drag and drop the crx file and add my extension but now I'm trying to do it programmatically. 

Where I neet to call the method LoadUnpackExtensionAsync reference above ? Actually when I call it after loading an url, I got an error : "DevTools Client Error :Method not available."

Even after add : 
```
   settings.CefCommandLineArgs.Add("remote-debugging-pipe", "1");
   settings.CefCommandLineArgs.Add("enable-unsafe-extension-debugging", "1");
```
And 

       context.SetPreference("extensions.ui.developer_mode", true, out errorString);

Can you explain me the best way to add and activate an extension in CefSharp without user interaction ? 

Link to the extension => https://chromewebstore.google.com/detail/ia4chrome/jifbnihciifbfeiiijegkfnbigagacjk

Thanks !

**Comment by mitchcapper** [comment]
You shouldn't need to explicitly be setting it as an expando object although master should handle the serialization of them now correctly over just the string/object dictionaries.

I didn't get the load unsafe extensions / devtools  call to work (got that "Method not available")  but per above I was able to throw the entire extension blob in there and get it to load a never before loaded extension without user interaction.  That only caveat being it caused a crash on first close as clearly it isn't meant to be loaded that way.   Now one could silently catch that first crash and afterwards it seems to behave perfectly.

Granted this was several CEF versions ago so things could be better now.

To test id recommend capturing the settings json, manually adding an extension, then capturing it again and getting the diff.

Next clear the session out startup get the extensions.settings pref, set the key, put it back and see if it is there under extensions.  Assuming it does continue working back to template it would for generating the ID and such to figure out where you are failing.

**Reply by max3163** [reply]
Thank you very much for your response ! 
The problem was the serialization which transform in the json an int to a float ( got 1 -> 1.0 ) and the plugin was not loading ! 
I create all class ( POCO ) and send the plain object to the context and it work ! ... but not in the first load ... Need to restart one time ...

I don't understand where/how to implement your example to work on the first time
```
protected override void OnRequestContextInitialized(IRequestContext requestContext)
{
    var curExtSettings = (Dictionary<string,object>)  requestContext.GetPreference("extensions.settings");
    curExtSettings["asdfasdfasdfasdf"] = Newtonsoft.Json.JsonConvert.DeserializeObject<IDictionary<string, object>>(extJson);
    if (! requestContext.SetPreference("extensions.settings",curExtSettings, out var err))
        throw new System.Exception(err);

    base.OnRequestContextInitialized(requestContext);
}
```
Any tips ?
